src/middleware/router.coffee

```diff
@@ -91,56 +94,73 @@ sendRequestToRoutes = (ctx, routes, next) ->
-            # handle mediator reponse
-            responseObj = JSON.parse response.body
-            ctx.mediatorResponse = responseObj
-            # then set koa response from responseObj.response
-            setKoaResponse ctx, responseObj.response
-          else
-            setKoaResponse ctx, response
-        .fail (reason) ->
-          # on failure
-          handleServerError ctx, reason
-      else
-        promise = buildNonPrimarySendRequestPromise ctx, route, options, path
-
-      promises.push promise
-
-    (Q.all promises).then ->
-      next()
-
+      do (route) ->
+        path = getDestinationPath route, ctx.path
+        options =
+          hostname: route.host
+          port: route.port
+          path: path
+          method: ctx.request.method
+          agent: false
+          rejectUnauthorized: true
+          key: keystore.key
+          cert: keystore.cert.data
+          secureProtocol: 'TLSv1_method'
+
+          headers: ctx.request.header
+        if route.cert?
+          options.ca = keystore.ca.id(route.cert).data
+
+        if ctx.request.querystring
+          options.path += '?' + ctx.request.querystring
+
+        if options.headers && options.headers.authorization
+          delete options.headers.authorization
+
+        if route.username and route.password
+          options.auth = route.username + ":" + route.password
+
+        if options.headers && options.headers.host
+          delete options.headers.host
+
+        if route.primary
+          promise = sendRequest(ctx, route, options)
+          .then (response) ->
+            logger.info "executing primary route : #{route.name}"
+            if response.headers?['content-type']?.indexOf('application/json+openhim') > -1
+              # handle mediator reponse
+              responseObj = JSON.parse response.body
+              ctx.mediatorResponse = responseObj
+              # then set koa response from responseObj.response
+              setKoaResponse ctx, responseObj.response
+            else
+              setKoaResponse ctx, response
+          .then ->
+            logger.info "primary route completed"
+            next()
+
+          .fail (reason) ->
+            # on failure
+            handleServerError ctx, reason
+        else
+          logger.info "executing non primary: #{route.name}"
+          nonPrimaryPromise = buildNonPrimarySendRequestPromise(ctx, route, options, path)
+          .then (response) ->
+            logger.info "Storing non primary route responses #{route.name}"
+            try
+              messageStore.storeNonPrimaryResponse ctx, response, ->
+                stats.incrementTransactionCount ctx, ->
+                stats.measureTransactionDuration ctx, ->
+            catch err
+              logger.error err
+
+        nonPrimaryPromises.push nonPrimaryPromise
+
+    (Q.all nonPrimaryPromises).then ->
+      logger.info "all non primary routes completed"
```

Comment:
This may be a better place to do the final transaction status update. Perhaps, create a new function in the message store that does the final update and call it here.